cmd/kubeapps-apis/core/packages/v1alpha1/packages_fan_in.go

```diff
@@ -0,0 +1,217 @@
+	summariesCh := make(chan summaryWithOffsets)
+
+	corePageSize := int(request.GetPaginationOptions().GetPageSize())
+	// We'll request a bit more than pageSize / n from each plugin. If the page
+	// size is 10 and we have 3 plugins, request 5 items from each to start.
+	pluginPageSize := corePageSize
+	if len(pkgPlugins) > 1 {
+		pluginPageSize = pluginPageSize / (len(pkgPlugins) - 1)
+	}
+
+	pluginPageOffsets := map[string]int{}
+	if request.GetPaginationOptions().GetPageToken() != "" {
+		err := json.Unmarshal([]byte(request.GetPaginationOptions().GetPageToken()), &pluginPageOffsets)
+		if err != nil {
+			return nil, fmt.Errorf("unable to unmarshal %q: %w", request.GetPaginationOptions().GetPageToken(), err)
+		}
+	}
+
+	fanInput := []<-chan *summaryWithOffset{}
+	for _, pluginWithSrv := range pkgPlugins {
+		// Importantly, each plugin needs its own request, with its on pagination
+		// options.
+		r := &packages.GetAvailablePackageSummariesRequest{
+			Context:       request.Context,
+			FilterOptions: request.FilterOptions,
+			PaginationOptions: &packages.PaginationOptions{
+				PageSize:  int32(pluginPageSize),
+				PageToken: fmt.Sprintf("%d", pluginPageOffsets[pluginWithSrv.plugin.Name]),
+			},
+		}
+
+		ch, err := sendAvailablePackageSummariesForPlugin(ctx, pluginWithSrv, r)
+		if err != nil {
+			return nil, err
+		}
+		fanInput = append(fanInput, ch)
+	}
+
+	// We now have a slice of channels for the fan-in and want a go routine that
+	// will ensure it sends the next (ordered) item from all channels down the
+	// channel.
+	go func() {
+		numSent := 0
+		nextItems := make([]*summaryWithOffset, len(fanInput))
+		for {
+			// Populate the empty next items from each channel.
+			for i, ch := range fanInput {
+				if nextItems[i] == nil {
+					// If the channel is closed, the value will remain nil.
+					ok := true
+					nextItems[i], ok = <-ch
+					if !ok {
+						// If the channel was closed, we reached the last item for that
+						// plugin. We need to recognise when all plugins have exhausted
+						// itemsoffsets
+						pluginName := pkgPlugins[i].plugin.Name
+						pluginPageOffsets[pluginName] = CompleteToken
+					}
+
+					if nextItems[i] != nil && nextItems[i].err != nil {
+						summariesCh <- summaryWithOffsets{
+							err: nextItems[i].err,
+						}
+						close(summariesCh)
+						return
+					}
+				}
+			}
+
+			// Choose the minimum by name and send it down the line.
+			// First find the first non-nil value as the min.
+			minIndex := -1
+			for i, s := range nextItems {
+				if s != nil {
+					minIndex = i
+					break
+				}
+			}
+
+			// If there is no non-nil value left, we're done.
+			if minIndex == -1 {
+				close(summariesCh)
+				return
+			}
+
+			// Otherwise, we find the minimum item of the next items from each channel.
+			for i, s := range nextItems {
+				if s != nil && s.availablePackageSummary.Name < nextItems[minIndex].availablePackageSummary.Name {
+					minIndex = i
+				}
+			}
```

Comment:
And this part find which item, out of the next items from each plugin, should be sent down the funnel.

Reply:
> In packages_fan_in.go you'll see that it is ensuring this is the case (and testing it). It's not a simple union, it's always grabbing the next minimum result (when comparing the next results of all plugins). I'll point out where...

AHHH, got it! Thanks for the explanation. I have missed that part; this is awesome! So each call will return an ordered collection of results (by name) as a result of the aggregation of each plugin's response.

Follow-up question: it works because we are assuming each plugin is returning its data with an order, no?  (I mean, if the response wasn't deterministic, this pagination approach wouldn't work at all, I guess)

Reply:
Correct. In particular, we're assuming that like the k8s end points, the plugins return results in `metadata.name` alpha order only. Unfortunately we can't provide any sorting (due to the restrictions on the k8s endpoints themselves) without reading the whole result and sorting or caching that, which I'm not planning to do for now :)